REGRESSION(274827@main): ASSERTION FAILED: !deletionHasBegun()

obyknovenius opened the pull request:
#### bbf9f2b123cb1366ff9bbe4932fd42f010dc6257
<pre>
REGRESSION(274827@main): ASSERTION FAILED: !deletionHasBegun()
<a href="https://bugs.webkit.org/show_bug.cgi?id=269784">https://bugs.webkit.org/show_bug.cgi?id=269784</a>

Reviewed by Chris Dumez.

Use `RefPtrAllowingPartiallyDestroyed` to reference the `Document`
in `LocalDOMWindow::removeAllEventListeners()` because it can be called
when the deletion of the `Document` has already begun.

* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::removeAllEventListeners):

Canonical link: <a href="https://commits.webkit.org/275098@main">https://commits.webkit.org/275098@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/091fbc96e9e3d203d9ea02e34973b8de457a229e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40870 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19883 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43248 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43428 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/36961 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/43177 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22885 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17214 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33871 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41444 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16825 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35231 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14492 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14594 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36221 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44717 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37079 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36532 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/40272 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15685 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12878 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38637 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17304 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17355 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5431 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16948 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2408,11 +2408,6 @@ void LocalDOMWindow::removeAllEventListeners()
 #endif
 
-#if ENABLE(TOUCH_EVENTS)
-    if (RefPtr document = this->document())
```
The less risky change would be to keep the code as-is but switch from `RefPtr` to `RefPtrAllowingPartiallyDestroyed<Document>`.

Your new code is not equivalent since `Document::removeAllEventListeners()` is NOT the only caller of `LocalDOMWindow::removeAllEventListeners()`.

obyknovenius replied:
Didn't know about `RefPtrAllowingPartiallyDestroyed`. Thank you for your advice.

cdumez commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -2409,7 +2409,7 @@ void LocalDOMWindow::removeAllEventListeners()
 
 #if ENABLE(TOUCH_EVENTS)
-    if (RefPtr document = this->document())
+    if (RefPtrAllowingPartiallyDestroyed document = this->document())
```
Note that I suggested `RefPtrAllowingPartiallyDestroyed<Document>`. This will NOT build.

obyknovenius replied:
Seems like it builds just fine. Should I still change it to `RefPtrAllowingPartiallyDestroyed<Document>`?

rniwa replied:
Yes, please change it. It builds because touch events aren't enabled on bots.

obyknovenius replied:
Done.

rniwa reviewed (changes requested):
Please make the change.

obyknovenius committed:
REGRESSION(274827@main): ASSERTION FAILED: !deletionHasBegun()
https://bugs.webkit.org/show_bug.cgi?id=269784

Reviewed by Chris Dumez.

Use `RefPtrAllowingPartiallyDestroyed` to reference the `Document`
in `LocalDOMWindow::removeAllEventListeners()` because it can be called
when the deletion of the `Document` has already begun.

* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::removeAllEventListeners):

Canonical link: https://commits.webkit.org/275098@main
